contextualized/dags/torch_notmad/torch_notmad.py

```diff
@@ -0,0 +1,190 @@
+            n_archetypes (int:4): Number of archetypes in explainer
+
+            Encoder Kwargs
+            ----------------
+            encoder_kwargs(dict): Dictionary of width, layers, and link_fn associated with encoder.
+
+            Optimization Kwargs
+            -------------------
+            learning_rate(float): Optimizer learning rate
+            opt_step(int): Optimizer step size
+
+            Loss Kwargs
+            -----------
+            sample_specific_loss_params (dict of str: int): Dict of params used by NOTEARS loss (l1, alpha, rho)
+            archetype_specific_loss_params (dict of str: int): Dict of params used by Archetype loss (l1, alpha, rho)
+
+        """
+        super(NOTMAD_model, self).__init__()
+
+        # dataset params
+        self.datamodule = datamodule
+        self.n_archetypes = n_archetypes
+        self.context_shape = self.datamodule.C_train.shape
+        self.feature_shape = self.datamodule.X_train.shape
+
+        # dag/loss params
+        self.project_distance = 0.1
+        self.archetype_loss_params = archetype_loss_params
+        self.alpha, self.rho, self.use_dynamic_alpha_rho = self._parse_alpha_rho(
+            sample_specific_loss_params
+        )
+
+        # model layer shapes
+        encoder_input_shape = (self.context_shape[1], 1)
+        encoder_output_shape = (self.n_archetypes,)
+        explainer_output_shape = (self.feature_shape[1], self.feature_shape[1])
+
+        # layer params
+        self.init_mat = init_mat
+
+        # opt params
+        self.learning_rate = learning_rate
+        self.opt_step = opt_step
+
+        # layers
+        self.encoder = NGAM(
+            encoder_input_shape[0],
+            encoder_output_shape[0],
+            layers=encoder_kwargs["layers"],
+            width=encoder_kwargs["width"],
+        )
+        self.explainer = Explainer(self.n_archetypes, explainer_output_shape)
+
+        # loss
+        self.my_loss = lambda x, y: self._build_arch_loss(
+            archetype_loss_params
+        ) + NOTEARS_loss(x, y, sample_specific_loss_params["l1"], self.alpha, self.rho)
+
+    def forward(self, c):
+        c = c.float()
+        c = self.encoder(c).float()
+        out = self.explainer(c)
+        return out.float()
+
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        sch = torch.optim.lr_scheduler.StepLR(
+            optimizer, step_size=self.opt_step, gamma=0.5
+        )
+        # learning rate scheduler
+        return {
+            "optimizer": optimizer,
+            "lr_scheduler": {
+                "scheduler": sch,
+                "monitor": "train_loss",
+            },
+        }
+
+    def training_step(self, batch, batch_idx):
+        C, x_true = batch
+        w_pred = self.forward(C).float()
+        loss = self.my_loss(x_true.float(), self._mask(w_pred)).float()
```

Comment:
It looks like `_mask` is only called during training, so this will output whatever the archetype diagonals were initialized with before training. We might want to set the archetype diagonals to 0 before training as well, you could do `explainer.set_archetypes(self._mask(explainer.get_archetypes()))` in `__init__`.